### DISCUSSION TITLE
Best way to skip to indexOf (BufferedSource.skipTo)?

### DISCUSSION THREAD

**Discussion by okiolover** [square/okio#1162] [Q&A]
I have a document that may or may not have a sequence of bytes I'm looking for. I don't care about the data before this sequence of bytes, but I do care about the data after this sequence of bytes, if it exists.

First, I thought to do this:
```kotlin
fun BufferedSource.skipTo(bytes: ByteString): Long {
  val index = indexOf(bytes)
  if (index == -1L) {
    skip(buffer.size)
  }
  return index
}
```

Unfortunately, indexOf expands the buffer and will even buffer my entire document in memory if the ByteString is not in the document, when I really just want to skip through the document.

To fix that, I wrote up my own skip function, checking eight kilobytes at a time:
```kotlin
fun BufferedSource.skipTo(bytes: ByteString): Long {
  while (true) {
    var index = buffer.indexOf(bytes)
    if (index != -1L) {
      skip(index)
      return index
    }
    buffer.clear()
    if (!request(8192L)) {
      // Last try.
      index = buffer.indexOf(bytes)
      if (index == -1L) {
        buffer.clear()
      } else {
        skip(index)
      }
      return index
    }
  }
}
```

Is there an easier way to do this with Okio?

**Comment by JakeWharton** [comment]
I would read one byte at a time until you match the first byte in your byte string. When you do, use rangeEquals to check if the remaining bytes match at the head of the source. If they do, you've found your march. If they don't, continue looping and reading one byte at a time.

Roughly:
```kotlin
fun BufferedSource.seekPast(bytes: ByteString) {
  require(bytes.isNotEmpty())
  val firstByte = bytes[0]
  while (true) {
    if (readByte() == firstByte && rangeEquals(0, bytes, 1, bytes.length - 1) {
      return true
    }
  }
}
```

Note: untested and written on mobile!

**Reply by okiolover** [reply]
That's a great idea; thank you. I'll post below what I got working.

**Comment by okiolover** [comment]
Thanks to Jake's suggestion, I used `rangeEquals`:
```kotlin
fun BufferedSource.skipTo(bytes: ByteString): Long {
  require(bytes.size != 0)
  val firstByte = bytes[0]
  var index = 0L
  while (true) {
    if (exhausted()) {
      return -1L
    }
    require(1)
    if (buffer[0] == firstByte && rangeEquals(0L, bytes, 1, bytes.size - 1)) {
      return index
    }
    skip(1)
    index++
  }
}
```
